src/wallet/test/coinselector_tests.cpp

```diff
@@ -453,6 +456,37 @@ BOOST_AUTO_TEST_CASE(bnb_search_test)
     }
 }
 
+BOOST_AUTO_TEST_CASE(tx_creation_bnb_sffo_restriction)
+{
+    // Verify the transaction creation process does not produce a BnB solution when SFFO is enabled.
+    // This is currently problematic because it could require a change output. And BnB is specialized on changeless solutions.
+    std::unique_ptr<CWallet> wallet = NewWallet(m_node);
+    WITH_LOCK(wallet->cs_wallet, wallet->SetLastBlockProcessed(300, uint256{})); // set a high block so internal UTXOs are selectable
+
+    CTxDestination change_dest = *Assert(wallet->GetNewChangeDestination(OutputType::BECH32));
+    CTxOut change_out(0, GetScriptForDestination(change_dest));
+    size_t change_output_size = GetSerializeSize(change_out);
+
+    CFeeRate effective_feerate(10000);
+    CFeeRate discard_feerate = GetDiscardRate(*wallet);
+
+    CAmount change_fee = effective_feerate.GetFee(change_output_size);
+    size_t change_spend_size = CalculateMaximumSignedInputSize(change_out, wallet.get(), nullptr);
+    CAmount cost_of_change = discard_feerate.GetFee(change_spend_size) + change_fee;
+
+    // Add spendable coin at the BnB selection upper bound
+    CoinsResult available_coins;
+    add_coin(available_coins, *wallet, COIN + cost_of_change, /*feerate=*/CFeeRate(0), /*nAge=*/6*24, /*fIsFromMe=*/true, /*nInput=*/0, /*spendable=*/true);
+
+    // Now verify coin selection does not produce BnB result
+    ASSERT_DEBUG_LOG_NOT_FOUND("Coin Selection: Algorithm:bnb");
+    CCoinControl coin_control;
+    coin_control.m_feerate = effective_feerate;
+    coin_control.destChange = change_dest;
+    CRecipient recipient{PubKeyDestination({}), /*target=*/COIN, /*subtract_fee=*/true};
+    BOOST_CHECK(CreateTransaction(*wallet, {recipient}, /*change_pos=*/-1, coin_control));
```

Comment:
In d62d49c605f89a325b929e8fb48091b0b56399ee "test: add coverage for BnB-SFFO restriction"

It seems like using `CreateTransaction` is a bit too high level for the goal of this test. Really this test is targeting the function that runs the various coin selection algorithms and chooses the result to return. This would be `ChooseSelectionResult`. Although this function is not exposed, its callers are, so this test could use either `SelectCoins` or `AutomaticCoinSelection` to test the same behavior. This would also remove the need to have the debug.log helper or even changing `CreatedTransactionResult` to contain the coin selection algorithm used since both of those functions return a `SelectionResult` which will contain the algo.

```suggestion
    FastRandomContext rand{};
    CoinSelectionParams params{
        rand,
        /*change_output_size=*/ 31,
        /*change_spend_size=*/ 68,
        /*min_change_target=*/ 0,
        /*effective_feerate=*/ CFeeRate(3000),
        /*long_term_feerate=*/ CFeeRate(1000),
        /*discard_feerate=*/ CFeeRate(1000),
        /*tx_noinputs_size=*/ 0,
        /*avoid_partial=*/ false,
    };
    params.m_subtract_fee_outputs = true;
    params.m_change_fee = params.m_effective_feerate.GetFee(params.change_output_size);
    params.m_cost_of_change = params.m_discard_feerate.GetFee(params.change_spend_size) + params.m_change_fee;

    // Add spendable coin at the BnB selection upper bound
    CoinsResult available_coins;
    add_coin(available_coins, *wallet, COIN + params.m_cost_of_change, /*feerate=*/CFeeRate(0), /*nAge=*/6*24, /*fIsFromMe=*/true, /*nInput=*/0, /*spendable=*/true);

    // Now verify coin selection does not produce BnB result
    auto result = SelectCoins(*wallet, available_coins, {}, COIN, {}, params);
    BOOST_CHECK(result.has_value());
    BOOST_CHECK(result->GetAlgo() != SelectionAlgorithm::BNB);
```

Reply:
Yeah. That would work as well.
The `CreateTransaction` usage comes from the previous version of this work (before we agreed to disable BnB when SFFO is enabled). Where BnB was enabled and this was verifying that the resulting transaction didn't contain a change output.

Reply:
Pulled.